### DISCUSSION TITLE
Airflow 2.2 removed Airflow group and moved into root group

### DISCUSSION THREAD

**Discussion by Alien2150** [apache/airflow#18904] [General]
### Apache Airflow version

2.2.0 (latest released)

### Operating System

Debian GNU/ Linux 10 (buster)

### Deployment

Other Docker-based deployment

### What happened

When I try to build my own images and run a COPY command it fails on 2.2 because the Airflow group is missing:

```
COPY --chown=airflow:airflow src/python/dags ./dags/python/
COPY --chown=airflow:airflow src/python/hooks ./hooks
COPY --chown=airflow:airflow src/python/operators ./operators/
COPY --chown=airflow:airflow src/sql/ ./dags/sql/
COPY --chown=airflow:airflow src/python/requirements.txt /tmp/requirements.txt
```
It seems the the airflow group was removed on 2.2.0 and now put into the root group. This is not a good security practice (imho). Is there a reason to have airflow running as root?

docker run --rm -it apache/airflow:2.2.0-python3.9 bash 
cat /etc/group: 
`root:x:0:airflow
daemon:x:1:
bin:x:2:
sys:x:3:
adm:x:4:
tty:x:5:
disk:x:6:
lp:x:7:
mail:x:8:
news:x:9:
uucp:x:10:
man:x:12:
proxy:x:13:
kmem:x:15:
dialout:x:20:
fax:x:21:
voice:x:22:
cdrom:x:24:
floppy:x:25:
tape:x:26:
sudo:x:27:
audio:x:29:
dip:x:30:
www-data:x:33:
backup:x:34:
operator:x:37:
list:x:38:
irc:x:39:
src:x:40:
gnats:x:41:
shadow:x:42:
utmp:x:43:
video:x:44:
sasl:x:45:
plugdev:x:46:
staff:x:50:
games:x:60:
users:x:100:
nogroup:x:65534:
ssh:x:101:`

docker run --rm -it apache/airflow:2.1.4-python3.9 cat /etc/group:
`
root:x:0:
daemon:x:1:
bin:x:2:
sys:x:3:
adm:x:4:
tty:x:5:
disk:x:6:
lp:x:7:
mail:x:8:
news:x:9:
uucp:x:10:
man:x:12:
proxy:x:13:
kmem:x:15:
dialout:x:20:
fax:x:21:
voice:x:22:
cdrom:x:24:
floppy:x:25:
tape:x:26:
sudo:x:27:
audio:x:29:
dip:x:30:
www-data:x:33:
backup:x:34:
operator:x:37:
list:x:38:
irc:x:39:
src:x:40:
gnats:x:41:
shadow:x:42:
utmp:x:43:
video:x:44:
sasl:x:45:
plugdev:x:46:
staff:x:50:
games:x:60:
users:x:100:
nogroup:x:65534:
ssh:x:101:
airflow:x:50000:airflow
`

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
> It seems the the airflow group was removed on 2.2.0 and now put into the root group. This is not a good security practice (imho). 

Why? What kind of security practice it vilolates? What problem it might cause? I'd love to hear some arguments for it. So far I had this discussion several times and no-one was able to say why this would be a bad practice (as oppose to running as `root` user which definitely is a bad practice and there are good arguments for it)..
But I am open to hear more arguments.

> Is there a reason to have airflow running as root?

Yes. That was a deliberate choice. Runnign as `root` group is recommended practice introduced by Open Shift (that's why I claim to say it is not a bad or dangerous practice) - this allows the image to run as arbitrary user as described here https://airflow.apache.org/docs/docker-stack/entrypoint.html#allowing-arbitrary-user-to-run-the-container

I am converting this into discussion, as I am eager to hear what are the arguments for `root` group being dangerous.

**Reply by potiuk** [reply]
And just to add - the original `airflow` group was there before, but It had not been used actually - in all the recommendations for a long time (> 6 months) we recommend to run all our images, and have all the docker-compose example and helm chart use GID=0 - precisely to maake it open-shift arbitrary-user compatible. This has the nice property that you can even run airflow with random user id (this is what Open Shift does by default) and everything will work fine.

**Reply by Alien2150** [reply]
Hey @potiuk, 

thank you for coming back to me that quickly. Thanks for sharing the link I will check it 🤗. I was simply not aware that it can be changed and probably this makes the discussion invalid. Of course I will share you a few thoughts on this one (Please correct me if I am wrong):

Adding to the root group means also being able to read/change "root only files" right? (Unless the group permission has changed on various files). So potentially I can read/change file-content that I should not be able to. Of course I would need some time to check various scenarios but some of I can think of (And maybe these are a bit exaggerated):

* I write a Dag to expose /var/logs/ - content to the UI / Logs
* Replace binary files by my own one (could be by-passed as group has ALL permissions):
  * Example: lrwxrwxrwx 1 root root       4 Sep 27 00:00 sh -> dash

**Reply by potiuk** [reply]
> Adding to the root group means also being able to read/change "root only files" right? (Unless the group permission has changed on various files). So potentially I can read/change file-content that I should not be able to. Of course I would need some time to check various scenarios but some of I can think of (And maybe these are a bit exaggerated):

Nope. You can read files that are "owned" by "root" group (and group readable). Not owned by "root" user. Those are different. In all the cases where "root" owned files are protected from reading, they have only "user" rights to read, group and other have "no-rights" by definition. 

Some example: `/etc/password` is "world readable" because in modern linux it only contains user names/ids/bash but no passwords (It used to contain encrypted password in the old dark days of UNIX). The passwords are kept in `/etc/shadow` - and yeah - this file is "group" readable but for a dedicated `shadow` group, not `root` group because the "password" validation process runs with "no-user/shadow group"  to not have too many permissions. Belonging to `root` group brings no "permissions" to read the "/etc/shadow" file.

```
root@dd453b36b5ed:/# ls -la /etc/passwd
-rw-r--r-- 1 root root 926 Oct 11 00:00 /etc/passwd
root@dd453b36b5ed:/# ls -la /etc/shadow
-rw-r----- 1 root shadow 501 Oct 11 00:00 /etc/shadow
root@dd453b36b5ed:/# 
```

Similarly you cannot replace binaries. Every file creaed by `root` is "owner" writable only. Any binary you have in the system that is "root" writable hase "user" writeable, "group" and "other" are "read-only".

The `umask` in debian based system is set in the way that any file creaated by "root" is by definition "group read" "owner write".
```
root@dd453b36b5ed:/# umask
0022
```
Via: https://unix.stackexchange.com/questions/84676/how-is-umask-calculated-in-linux

> umask 0022 would make the new mask 0644 (0666-0022=0644) meaning that group and others have read (no write or execute) permissions.
> The "extra" digit (the first number = 0), specifies that there are no special modes.
> If mode begins with a digit it will be interpreted as octal otherwise its meant to be symbolic

 We only change the umask later in our image for the logged-in user (in the entrypoint), so that whenever "airflow" creates files, they are "group writeable" - but this happens only after the image is build and ready and when 'the logged-in user" executes the entrypoint. By definition umask 

"Root" group historically had no more permission or access than any other group. No extra permissions, no extra

**Reply by potiuk** [reply]
`Example: lrwxrwxrwx 1 root root 4 Sep 27 00:00 sh -> dash`

This is not permission of the file. It's always like that for symbolic link. You cannot replace the link unless you have `wx` permission to the directory where the link is located (which `root` group does not have). Nor you cannot write to the file link points to (unless you have write permission for the target file). 

Try it.

**Reply by potiuk** [reply]
* `I write a Dag to expose /var/logs/ - content to the UI / Logs`

This is precisely what we want. We want any user to be able to write to logs, so that any other user who is in the `root` group can both read and write. This is precisely what "allow arbitrary user to run the image" is about. We WANT every file created by the user who runs "airflow" to be accessible by any other user who belongs to the `root` group. In OpenShift for example it works in the way that the same "worker" can be restarted and it can have a different, random user. That user should be able to act in the same way as any other user and read/write/delete the files created by the other users who are in the `root` group. This is why we deliberately set the umask in the entrypoint:

```
# Set umask to 0002 to make all the directories created by the current user group-writeable
# This allows the same directories to be writeable for any arbitrary user the image will be
# run with, when the directory is created on a mounted volume and when that volume is later
# reused with a different UID (but with GID=0)
umask 0002
```
So this is the Open-Shift compatibility "feature", not a security issue.
